test/Npgsql.Tests/PoolTests.cs

```diff
@@ -39,6 +40,97 @@ public void MinPoolSizeLargerThanMaxPoolSize()
             Assert.That(() => CreateConnection(connString), Throws.Exception.TypeOf<ArgumentException>());
         }
 
+        [Test]
+        [TestCase(true)]
+        [TestCase(false)]
+        public async Task DoesClearPoolWhenDetectingNetworkOrFatalBackendError(bool terminateBackendInsteadOfRecoverableError)
+        {
+            // Arrange
+            // Create two separate connection strings in order to have multiple pools for testing
+            var connString = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                ApplicationName = nameof(DoesClearPoolWhenDetectingNetworkOrFatalBackendError)
+            }.ToString();
+
+            var secondConnString = new NpgsqlConnectionStringBuilder(ConnectionString)
+            {
+                ApplicationName = $"{nameof(DoesClearPoolWhenDetectingNetworkOrFatalBackendError)}:2"
+            }.ToString();
+
+            // Act
+            using (var conn = CreateConnection(connString))
+            using (var conn2 = CreateConnection(connString))
+            {
+                // Open Two connections, noting the Backend Ids
+                conn.Open();
+                conn2.Open();
+
+                // Note connector backend Ids
+                var backendIds = new[]
+                {
+                    conn.Connector!.BackendProcessId,
+                    conn2.Connector!.BackendProcessId
+                };
+
+                // Issue a (syntactically invalid) query in order to invoke a reader exception that is not fatal
+                // This is to ensure we do not clear the pool upon regular errors (that is not desirable)
+                foreach(var c in new[] { conn, conn2})
+                    Assert.ThrowsAsync<PostgresException>(async () => await c.ExecuteScalarAsync("this_query_is_invalid_on_purpose"));
+
+                // Return connectors to pool
+                await conn.CloseAsync();
+                await conn2.CloseAsync();
+
+                // Terminate the underlying backend processes to break the Connector connections from our pool
+                if (terminateBackendInsteadOfRecoverableError) {
+                    using (var conn3 = CreateConnection(secondConnString)) {
+                        conn3.Open();
+
+                        foreach (var backendId in backendIds)
+                            await conn3.ExecuteNonQueryAsync($"SELECT pg_terminate_backend({backendId})");
+                    }
+
+                    // Wait for backends to be terminated
+                    await Task.Delay(1000);
```

Comment:
Flakiness potential

Reply:
I've replaced this with a new method, `TerminateBackendAndWaitAsync`, which will terminate and check for termination for a few seconds (or fail) before proceeding.